### ISSUE TITLE
A bug for using KEY_SET method to deliver the message to swss & orchagent

### ISSUE THREAD

**Issue by guking3** [Azure/sonic-swss-common#493]
Hello fellows and everybody:
     I founld a bug in the KEY_SET method to deliver the message as belove:
#### Description
When using the  KEY_SET method to deliver the message to swss&orchagent, ConsumerStateTable judge the msg DEL or SET by getting the value of the KEY_SET.Under normal condition, the SET msg can deliver the key to the KEY_SET, then SET the value with the key in hash table(or others), the swss-comm ConsumerSateTable::pops process as belove
RedisReply r(m_db, command);
    auto ctx0 = r.getContext();
    vkco.clear();

    // if the set is empty, return an empty kco object
    if (ctx0->type == REDIS_REPLY_NIL)
    {
        return;
    }

    assert(ctx0->type == REDIS_REPLY_ARRAY);
    size_t n = ctx0->elements;
    vkco.resize(n);
    for (size_t ie = 0; ie < n; ie++)
    {
        auto& kco = vkco[ie];
        auto& values = kfvFieldsValues(kco);
        assert(values.empty());

        auto& ctx = ctx0->element[ie];
        assert(ctx->element[0]->type == REDIS_REPLY_STRING);
        std::string key = ctx->element[0]->str;
        kfvKey(kco) = key;

        assert(ctx->element[1]->type == REDIS_REPLY_ARRAY);
        auto ctx1 = ctx->element[1];
        for (size_t i = 0; i < ctx1->elements / 2; i++)
        {
            FieldValueTuple e;
            fvField(e) = ctx1->element[i * 2]->str;
            fvValue(e) = ctx1->element[i * 2 + 1]->str;
            values.push_back(e);
        }

        // if there is no field-value pair, the key is already deleted
        if (values.empty())
        {
            kfvOp(kco) = DEL_COMMAND;
        }
        else
        {
            kfvOp(kco) = SET_COMMAND;
        }
the value can be get in the table,so the operator is SET_COMMAND
If the msg is DEL, ConsumerStateTable::pops get no value of the key ,the operator is DEL_COMMAND 

But, the question is that if the was a SET msg after DEL msg, the value was set in the hash table with the same key,
the r.getContext would got the value, it could not deliver the msg to the Business processing （Sometime， the DEL msg should must be deliver）

#### Steps to reproduce the issue:
1.LuaScript:DEL [INTF_TABLE_CHANNEL INTF_TABLE_KEY_SET _INTF_TABLE:ep195.1 INTF_TABLE_DEL_SET] [G ep195.1]
2.LuaScript:SET [INTF_TABLE_CHANNEL INTF_TABLE_KEY_SET _INTF_TABLE:ep195.1 _INTF_TABLE:ep195.1 _INTF_TABLE:ep195.1 _INTF_TABLE:ep195.1 _INTF_TABLE:ep195.1] [G ep195.1 admin_status up switch_port disable mac  mtu 9100 vrf_name ]
3.Two msg publish within 1 Milliseconds

#### Describe the results you received:
When got the publish msg of the DEL in INTF_TABLE:ep195.1 ,the ConsumerStateTable::pops msg would judge it as SET_COMMAND, because the next SET msg had aleady set the value to the key of  INTF_TABLE:ep195.1
Thus, the DEL msg could not be found in the ConsumerStateTable::pops, which would be very necessary for the orchangent process using.
#### Describe the results you expected:
Got the DEL msg in use

**Comment by guking3** [comment]
Hello fellows,
          I found this problem by testing the rapid distribution of interface VRF. During processing the VRF attribute setting of the  port(for example port ep195.1 ), the system has a default VRF name of default. When the interface is created without binding the related VRF, the VRF default setting is adopted by default. When the interface is bound to a configured VRF instance, such as Vrf_new. At this time, SWSS common needs to judge whether to add or delete messages by  swss-comm ConsumerSateTable::pops.
The message flow is as follows:
1.INTF_PORT ep195.1 DEL
2.INTF_PORT ep195.1 SET vrf Vrf_new

Using Lua script publishing KEY_SET method
1.luaDel SCRIPT
luaDel := redis.NewScript(`
		redis.call('SADD', KEYS[2], ARGV[2])
		redis.call('SADD', KEYS[4], ARGV[2])
		redis.call('DEL', KEYS[3])
		redis.call('PUBLISH', KEYS[1], ARGV[1]) 
		`)

2.luaSet SCRIPT
luaSet := redis.NewScript(`
		redis.call('SADD', KEYS[2], ARGV[2])
		for i = 0, #KEYS - 3 do
			redis.call('HSET', KEYS[3 + i], ARGV[3 + i * 2], ARGV[4 + i * 2])
		end
		redis.call('PUBLISH', KEYS[1], ARGV[1]) 
		`)

Both luaDel and luaSet have the same key ep155.1 ,then problems may occur. First got the  notification of subscriber msg del,but when get the value with the key ep155.1 the table had already set to vaule because of luaSet.So the DEL msg can not be distinguish.

In short, in the sde process of business code, VRF will be brought after creating interface. That is to say, a VRF must be brought when creating RIF. Now the interface can exist first, and then configure VRF. At this time, we must delete RIF first, and then create RIF with VRF. Therefore, the deletion message must be identified and distributed to the business processing.
            Best Regards，and looking forward to your discussion, communication and reply. Thank you!

**Comment by guking3** [comment]
Hello fellows,
	A simple test can be used to identify the problem as below:
-----Condition 1-------
Only DEL msg publish for ConsumerStateTable::pops

root@sonic:~# redis-cli 
127.0.0.1:6379> SADD INTF_TABLE_KEY_SET ep195.1
(integer) 1
127.0.0.1:6379> SADD INTF_TABLE_DEL_SET ep195.1
(integer) 1
127.0.0.1:6379> DEL _INTF_TABLE:ep195.1
(integer) 0
127.0.0.1:6379> PUBLISH INTF_TABLE_CHANNEL G
(integer) 0

ConsumerStateTable::pops with the key ep195.1,and find nill value of this key ep155.1, take it as DEL msg

127.0.0.1:6379> EVALSHA 88270a7c5c90583e56425aca8af8a4b8c39fe757 3 INTF_TABLE_KEY_SET INTF_TABLE: INTF_TABLE_DEL_SET 8192 _
1) 1) "ep155.1"
   2) (empty array)
   
No problem

-----Condition 2-------
Only SET msg publish for ConsumerStateTable::pops

root@sonic:~# redis-cli 
127.0.0.1:6379> SADD INTF_TABLE_KEY_SET ep195.1
(integer) 0
127.0.0.1:6379> HSET _INTF_TABLE:ep195.1 vrf_name Vrf_new
(integer) 1
127.0.0.1:6379> PUBLISH INTF_TABLE_CHANNEL G
(integer) 0

ConsumerStateTable::pops with the key ep195.1,and find value of this key ep155.1, take it as SET msg

127.0.0.1:6379> EVALSHA 88270a7c5c90583e56425aca8af8a4b8c39fe757 3 INTF_TABLE_KEY_SET INTF_TABLE: INTF_TABLE_DEL_SET 8192 _
1) 1) "ep195.1"
   2) 1) "mtu"
      2) "9100"
   
No problem

-----Condition 3-------
A SET msg after DEL msg publish immediately for ConsumerStateTable::pops with the same key

root@sonic:~# redis-cli
127.0.0.1:6379> SADD INTF_TABLE_KEY_SET ep155.1
(integer) 0
127.0.0.1:6379> SADD INTF_TABLE_DEL_SET ep155.1
(integer) 1
127.0.0.1:6379> DEL _INTF_TABLE:ep155.1
(integer) 0
127.0.0.1:6379> PUBLISH INTF_TABLE_CHANNEL G
(integer) 0
127.0.0.1:6379> 
127.0.0.1:6379> SADD INTF_TABLE_KEY_SET ep155.1
(integer) 0
127.0.0.1:6379> HSET _INTF_TABLE:ep155.1 vrf_name Vrf_new
(integer) 1
127.0.0.1:6379> PUBLISH INTF_TABLE_CHANNEL G
(integer) 0
127.0.0.1:6379> EVALSHA 88270a7c5c90583e56425aca8af8a4b8c39fe757 3 INTF_TABLE_KEY_SET INTF_TABLE: INTF_TABLE_DEL_SET 8192 _
1) 1) "ep155.1"
   2) 1) "mtu"
      2) "9100"
127.0.0.1:6379> 

ConsumerStateTable::pops got only one msg and take it as a SET msg,the DEL msg was lost.

The problem happened.


PS.
Can both of the two msg be got by the ConsumerStateTable::pops method with the same key,such as
1) 1) "ep155.1"
   2) (empty array)
2) 1) "ep155.1"
   2) 1) "mtu"
      2) "9100"

**Comment by guking3** [comment]
@qiluo-msft

**Comment by qiluo-msft** [comment]
Thanks! We are looking into this issue.

**Comment by qiluo-msft** [comment]
@prsunny to check VRF DEL/SET workflow

**Comment by guking3** [comment]
Dear experts，are there further developments about this issue? @qiluo-msft @prsunny 
                                                                                  Best Regards，Thank you!

**Comment by qiluo-msft** [comment]
We are still discussion internally. Will update soon.

**Comment by qiluo-msft** [comment]
Hi @guking3,

@prsunny, @lguohan and I discussed and propose modifying IntfsOrch to handle this case.

We understand SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID is a create_only attribute and it is a limitation of SAI. However orchagent could translate the ApplDb operation into workable SAI operations, ie. translate the merged SET operation (which has a different vrf compared with previous attribute) by SAI remove and create function calls.
